README.md
# [cookie-encrypter](https://github.com/ebourmalo/cookie-encrypter)

[![npm](https://img.shields.io/npm/v/cookie-encrypter.svg)](https://www.npmjs.com/package/cookie-encrypter)
[![npm](https://img.shields.io/npm/dm/cookie-encrypter.svg)](https://www.npmjs.com/package/cookie-encrypter)

Transparently encrypt/decrypt your cookie using an express middleware to set after the  [cookie-parser](https://github.com/expressjs/cookie-parser).
Support all type of cookie (including http-only and signed) with string content or JSON.
Use `aes256` as the default encryption algorithm (internally use the nodejs [crypto](https://nodejs.org/api/crypto.html) module).

## Installation

```sh
$ npm install cookie-encrypter
```

## Example

Easy to use:
```js
const express = require('express')
const cookieParser = require('cookie-parser')
const cookieEncrypter = require('cookie-encrypter')
// we use a 32bits long secret key (with aes256)
const secretKey = 'foobarbaz1234567foobarbaz1234567'
const cookieParams = {
  httpOnly: true,
  signed: true,
  maxAge: 300000
}

const app = express()
app.use(cookieParser(secretKey))
// use it as a simple middleware
app.use(cookieEncrypter(secretKey))

app.get('/setcookies', (req, res) => {
  // Set encrypted cookies
  res.cookie('supercookie', 'my text is encrypted', cookieParams)
  res.cookie('supercookie2', { myData: 'is encrypted' }, cookieParams)

  // You can still set plain cookies
  res.cookie('plaincookie', 'my text is plain', { plain: true })
  res.cookie('plaincookie2', { myData: 'is plain' }, { plain: true })

  res.json({ status: 'updated' })
})

app.get('/getcookies', (req, res) => {
  console.log('Decrypted cookies: ', req.signedCookies)
  console.log('Plain cookies: ', req.cookies)

  res.json({ status: 'ok' })
})

app.listen(8080)
```

[You can find a ready-to-use example here](https://github.com/ebourmalo/cookie-encrypter/tree/master/example)
Think about the `npm install` before running it ;)

## API

### cookieEncrypter(secret, options)

- `secret` a string or array used for encrypting cookies.
- `options` an optional object to set options for encryption.
-  `options.algorithm` algorithm used to encrypt cookie data (any algorithm supported by OpenSSL). `aes256` used as the default one.

### cookieEncrypter.encryptCookie(str, options)

Encrypt a cookie value and return it. An `options.algorithm` can optionaly be passed to specify an algorithm to use for the encryption.

### cookieEncrypter.decryptCookie(str, options)

Decrypt a cookie value and return it. An `options.algorithm` can optionaly be passed to specify an algorithm to use for the decryption.


## CHANGELOG

[See the changelog](https://github.com/ebourmalo/cookie-encrypter/blob/master/CHANGELOG.md)


CHANGELOG.md
v1.0.1 / 2018-05-29
===================

  * Use initialization vector for encryption to improve security [#5](https://github.com/ebourmalo/cookie-encrypter/commit/31c9078676687eedde4018a75b831cf248f6fb7d)
  * Add checks and throw errors at initialisation if wrong options are passed
  * Update code to ES6

v0.2.3 / 2016-11-18
===================

  * Update README and add example for both cases (encrypted cookies and plain cookies)

v0.2.0 / 2016-11-18
===================

  * Add an option to `res.cookie` to set plain cookie (without encryption). Thanks to @skarbovskiy for [his PR](https://github.com/ebourmalo/cookie-encrypter/pull/2)


index.test.js
const crypto = require('crypto')
const { decryptCookie, encryptCookie } = require('./index')

describe('encryptCookie', () => {

  function testEncyryptCookie(algorithm, keyLengthBytes) {
    const randomArrayString = Array.from({length: 40}, () => crypto.randomBytes(Math.floor(1 + Math.random() * 1000)).toString('hex'))
    randomArrayString.forEach(strToEncrypt => {
      it(`can encode an string with a ${keyLengthBytes} bytes key`, () => {
        const key = crypto.randomBytes(keyLengthBytes)
        expect(typeof encryptCookie(strToEncrypt, { algorithm, key })).toBe('string')
      })
    })

    const randomArrayBytesNumber = Array.from({length: 40}, () => Math.floor(1 + Math.random() * 1000)).filter(el => el !== keyLengthBytes)
    randomArrayBytesNumber.forEach(nbBytes => {
      it(`throw with a ${nbBytes} bytes key`, () => {
        const stringToEncrypt = 'StringToEncrypt'
        const key = crypto.randomBytes(nbBytes)
      
        expect(() => encryptCookie(stringToEncrypt, { algorithm, key })).toThrow()
      })
    })
  }

  describe('aes256 algorithm', () => {
    const algorithm = 'aes256' 
    const keyLengthBytes = 32
    testEncyryptCookie(algorithm, keyLengthBytes)
  })

  describe('aes-128-cbc algorithm', () => {
    const algorithm = 'aes-128-cbc' 
    const keyLengthBytes = 16
    testEncyryptCookie(algorithm, keyLengthBytes)
  })
})

describe('decryptCookie', () => {
  function testDecryptCookie(algorithm, keyLengthBytes) {
    it('can decode a string while using the good key', () => {
      const originalStr = crypto.randomBytes(Math.floor(1 + Math.random() * 1000)).toString('hex')
      const key = crypto.randomBytes(keyLengthBytes)
      const encyptedStr = encryptCookie(originalStr, { algorithm, key })

      expect(decryptCookie(encyptedStr, { algorithm, key })).toEqual(originalStr)
    })

    it('throw an error while decrypting using a wrong key', () => {
      const originalStr = crypto.randomBytes(Math.floor(1 + Math.random() * 1000)).toString('hex')
      const key = crypto.randomBytes(keyLengthBytes)
      const wrongKey = crypto.randomBytes(keyLengthBytes)
      const encyptedStr = encryptCookie(originalStr, { algorithm, key })

      expect(() => decryptCookie(encyptedStr, { algorithm, key: wrongKey })).toThrow()
    })

    it('throw an error while decrypting using a wrong-length key', () => {
      const originalStr = crypto.randomBytes(Math.floor(1 + Math.random() * 1000)).toString('hex')
      // use a different length
      const key = crypto.randomBytes(keyLengthBytes)
      const wrongSizeKey = key.toString('hex').substr(2)
      const encyptedStr = encryptCookie(originalStr, { algorithm, key })

      expect(() => decryptCookie(encyptedStr, { algorithm, key: wrongKey })).toThrow()
    })
  }

  describe('aes256 algorithm', () => {
    const algorithm = 'aes256' 
    const keyLengthBytes = 32
    
    testDecryptCookie(algorithm, keyLengthBytes)
  })

  describe('aes-128-cbc algorithm', () => {
    const algorithm = 'aes-128-cbc' 
    const keyLengthBytes = 16
    
    testDecryptCookie(algorithm, keyLengthBytes)
  })
})


example/index.js
const express = require('express')
const cookieParser = require('cookie-parser')
const cookieEncrypter = require('../index')
// we use a 32bits long secret key (with aes256)
const secretKey = 'foobarbaz1234567foobarbaz1234567'
const cookieParams = {
  httpOnly: true,
  signed: true,
  maxAge: 300000
}

const app = express()
app.use(cookieParser(secretKey))
// use it as a simple middleware
app.use(cookieEncrypter(secretKey))

app.get('/setcookies', (req, res) => {
  // Set encrypted cookies
  res.cookie('supercookie', 'my text is encrypted', cookieParams)
  res.cookie('supercookie2', { myData: 'is encrypted' }, cookieParams)

  // You can still set plain cookies
  res.cookie('plaincookie', 'my text is plain', { plain: true })
  res.cookie('plaincookie2', { myData: 'is plain' }, { plain: true })

  res.json({ status: 'updated' })
})

app.get('/getcookies', (req, res) => {
  console.log('Decrypted cookies: ', req.signedCookies)
  console.log('Plain cookies: ', req.cookies)

  res.json({
    encryptedCookies: req.signedCookies,
    plainCookies: req.cookies    
  })
})

app.listen(8080)
